cmd/snap/cmd_routine_portal_info.go

```diff
@@ -0,0 +1,145 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"path/filepath"
+	"text/template"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/sandbox/apparmor"
+	"github.com/snapcore/snapd/sandbox/cgroup"
+)
+
+type cmdRoutinePortalInfo struct {
+	clientMixin
+	PortalInfoOptions struct {
+		Pid int
+	} `positional-args:"true" required:"true"`
+}
+
+var shortRoutinePortalInfoHelp = i18n.G("Return information about a process")
+var longRoutinePortalInfoHelp = i18n.G(`
+The portal-info command returns information about a process in keyfile format.
+
+This command is used by the xdg-desktop-portal service to retrieve
+information about snap confined processes.
+`)
+
+func init() {
+	addRoutineCommand("portal-info", shortRoutinePortalInfoHelp, longRoutinePortalInfoHelp, func() flags.Commander {
+		return &cmdRoutinePortalInfo{}
+	}, nil, []argDesc{{
+		// TRANSLATORS: This needs to begin with < and end with >
+		name: i18n.G("<process ID>"),
+		// TRANSLATORS: This should not start with a lowercase letter.
+		desc: i18n.G("Process ID of confined app"),
+	}})
+}
+
+const portalInfoTemplate = `[Snap Info]
+InstanceName={{.Snap.Name}}
+{{- if .App}}
+AppName={{.App.Name}}
+{{- end}}
+{{- if .DesktopFile}}
+DesktopFile={{.DesktopFile}}
+{{- end}}
+HasNetwork={{.HasNetwork}}
+`
+
+var (
+	cgroupSnapNameFromPid   = cgroup.SnapNameFromPid
+	apparmorSnapNameFromPid = apparmor.SnapNameFromPid
+)
+
+func (x *cmdRoutinePortalInfo) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	snapName, err := cgroupSnapNameFromPid(x.PortalInfoOptions.Pid)
+	if err != nil {
+		return err
+	}
+	snap, _, err := x.client.Snap(snapName)
+	if err != nil {
+		return fmt.Errorf("cannot retrieve info for snap %q: %v", snapName, err)
+	}
+
+	// Try to identify the application name from AppArmor
```

Comment:
is this one bit you are unhappy with?

Reply:
If I could get the application information in a way that worked on the non-AppArmor platforms, that would be great.  But the fallback code (picking the first app that provides a desktop file) is going to give a reasonable result for most snaps: either the snap only contains a single desktop file, or the user will recognise the desktop files as being related to the same package.

Reply:
I think the coming code will allow for that directly from the cgroup data but we can change this when we get there